packages/wasm/crate/src/planner.rs

```diff
@@ -1,591 +1,389 @@
-        // for the cost of any additional `Spend` and `Output` actions necessary to pay the fee,
-        // we need to now calculate the transaction's fee again and capture the excess as change
-        // by subtracting the excess from the required value balance.
-        //
-        // Here, tx_real_fee is the minimum fee to be paid for the transaction, with no tip.
-        let mut tx_real_fee = self.gas_prices.fee(&self.plan.gas_cost());
-
-        // Since the excess fee paid will create an additional Output action, we need to
-        // account for the necessary fee for that action as well.
-        tx_real_fee += self.gas_prices.fee(&gas::output_gas_cost());
-
-        // For any remaining provided balance, add the necessary fee for collecting:
-        tx_real_fee += Amount::from(self.balance.provided().count() as u64)
-            * self.gas_prices.fee(&gas::output_gas_cost());
-
-        // Apply the fee tier to tx_real_fee so the block proposer can receive a tip:
-        tx_real_fee = Fee::from_staking_token_amount(tx_real_fee)
-            .apply_tier(self.fee_tier)
-            .amount();
-
-        assert!(
-            tx_real_fee <= self.plan.transaction_parameters.fee.amount(),
-            "tx real fee {:?} must be less than planned fee {:?}",
-            tx_real_fee,
-            self.plan.transaction_parameters.fee.amount()
-        );
-        let excess_fee_spent = self.plan.transaction_parameters.fee.amount() - tx_real_fee;
-        self.balance += Value {
-            amount: excess_fee_spent,
-            asset_id: *STAKING_TOKEN_ASSET_ID,
+    let mut iterations = 0usize;
+
+    while let Some(required) = actions
+        .balance_with_fee_estimate(&gas_prices, &fee_tier)
+        .required()
+        .next()
+    {
+        // Spend a single note towards the required balance, if possible.
+        let Some(note) = notes_by_asset_id
+            .get_mut(&required.asset_id)
+            .expect("we already queried")
+            .pop()
+        else {
+            return Err(anyhow!(
+                "ran out of notes to spend while planning transaction, need {} of asset {}",
+                required.amount,
+                required.asset_id,
+            )
+            .into());
         };
-        self.plan.transaction_parameters.fee = Fee::from_staking_token_amount(tx_real_fee);
+        actions.push(SpendPlan::new(&mut OsRng, note.note, note.position).into());
 
-        // For any remaining provided balance, make a single change note for each
-        for value in self.balance.provided().collect::<Vec<_>>() {
-            self.output(value, self_address);
-        }
+        // Recompute the change outputs, without accounting for fees.
+        actions.refresh_change(change_address);
+        // Now re-estimate the fee of the updated transaction and adjust the change if possible.
+        let fee = actions.fee_estimate(&gas_prices, &fee_tier);
+        actions.adjust_change_for_fee(fee);
```

Comment:
This first updates all the planned change outputs, so they're in place and counting towards the gas and hence the fee estimate, and _then_ adjusts the change note for the staking token to fund as much of the fee estimate as possible. 

So, if we've spent more than is required, we compute all the necessary change, and then "spend out of the change note" to cover fees.

Or, if we've spent less than is required, we'll set the value of the change output to 0, so we have a smaller imbalance to try to repair in the next iteration of the loop.